### ISSUE TITLE
Update Debugging Docs

### ISSUE THREAD

**Issue by ghost** [reactioncommerce/reaction-docs#962]
https://docs.reactioncommerce.com/docs/testing-debugging-server-code

This document is woefully out of date and need to be refreshed so users can set-up a step-debugger for Node within a Docker container. This issue should supersede https://github.com/reactioncommerce/reaction-docs/issues/854.

**Comment by ghost** [comment]
If you're debugging the API check `package.json` in the `reaction` repo for existing NPM/Yarn scripts for use in debugging such as `inspect:docker`. Run those like this assuming `dcr` is aliased to `docker-compose run --rm`:

```sh
dcr -p 3000:3000 -p 9229:9229 api npm run inspect:docker
```

To set up step VSCode debugging using `reaction-development-platform` to debug `example-storefront` as an example:

1. Create a `.vscode/launch.json` file in the `reaction-development-platform` root with the following contents:
    ```json
    {
      "version": "0.2.0",
      "configurations": [
        {
          "type": "node",
          "request": "attach",
          "name": "Docker: Attach to Example Storefront",
          "port": 9229,
          "address": "localhost",
          "localRoot": "${workspaceFolder}/example-storefront",
          "remoteRoot": "/usr/local/src/app",
          "protocol": "inspector",
          "skipFiles": [
            "<node_internals>/**"
          ]
        }
      ]
    }
    ```
2. Copy the `example-storefront/bin/start` script to create a debugging script like:
    ```sh
    cd example-storefront/bin && \
    cp start debug && \
    chmod +x debug
    ```
3. Copy the `start:dev` in `example-storefront/package.json` to create a `start:debug` NPM script.
4. Update the `start:debug` script created and prepend `NODE_OPTIONS='--inspect-brk=0.0.0.0:9229'` to the command.
5. Modify the `bin/debug` file created earlier and change the part that says `start:dev` to `start:debug`.
6. Modify `example-storefront/docker-compose.dev.yml` (once you `make dev-link-example-storefront` to create it) and modify it to add `ports` and change the `command` used, leaving `image` whatever it was before:
    ```
    services:
      web:
        ports:
          - "4000:4000"
          - "9229:9229"
        image: reactioncommerce/node-dev:12.14.1-v1
        # command: "./bin/start"
        command: "./bin/debug"
    ```
7. Restart the example storefront docker container and observe the output. You should see output like:
    ````
    web_1  | yarn run v1.21.1
    web_1  | $ NODE_ENV=development NODE_OPTIONS='--inspect-brk:0.0.0.0:9229' node ./src/server.js
    web_1  | Debugger listening on ws://127.0.0.1:9229/4342dc08-500d-4887-87c4-a26ab04bbf69
    web_1  | For help, see: https://nodejs.org/en/docs/inspector
    web_1  | Running NextJS server in development mode...
    ````
8. If port `9229` is used by an existing process find out what process it is and why it's using it. For example, it [may be being used by Admin](https://github.com/reactioncommerce/reaction-admin/commit/e6ca87be8cf4c9291f9af75d56a81a5bac4a5a90). You can shut off or change port of you like but you should know why it was being used. You could also update the `launch.json` and the `package.json` to use a different port.
9. Once you running example storefront in debug mode press `F5` in VSCode to start the debugger. The debugger should break on the first line of the application. From there you can set a breakpoint where you want to debug and then resume the debugger and take whatever action is necessary on the actual storefront in a browser to hit that breakpoint.

**Note:** You may prefer to create a new `docker-compose.debug.yml` file for the purpose of debugging so you can symlink that as the override for a given repo and develop multiple repos at a time yet leave only one with an open port `9229` for debugging in vscode.

Here are some additional resources:

- https://code.visualstudio.com/Docs/editor/debugging
- https://docs.microsoft.com/en-us/visualstudio/debugger/attach-to-running-processes-with-the-visual-studio-debugger?view=vs-2019#BKMK_Troubleshoot_attach_errors
- https://blog.risingstack.com/how-to-debug-a-node-js-app-in-a-docker-container/
- https://blog.theodo.com/2018/10/setting-debugger-node-js-docker-vs-code/
- https://medium.com/the-node-js-collection/node-options-has-landed-in-8-x-5fba57af703d

**Comment by ghost** [comment]
Just created a more fun tutorial with pictures: https://habd.as/post/reaction-commerce-debugging-primer/ Enjoy!
